blogposts/2020/optimizing-a-code-intelligence-indexer.md

```diff
@@ -0,0 +1,196 @@
+
+#### Use the compute that's ready for the taking
+
+The implementation of the indexer has historically been single-threaded for simplicity.
+
+![All my idle cores](https://i.imgflip.com/4bwccv.jpg)
+
+This design decision has recently become a heavier weight on our shoulders ever since we noticed one of our enterprise customers give their engineers [m5.24xlarge EC2 instances](https://aws.amazon.com/ec2/instance-types/).
+
+There's no way we're going to miss the opportunity to see all 96 of those cores light up the next time we chat. So we made everything that can run at the same time [run at the same time](https://github.com/sourcegraph/lsif-go/pull/91). Unfortunately, parallelizing workloads in general is not a trivial task.
+
+##### You have to protect shared state
+
+We did this by introducing a mutex around each shared map and slice that can be read from and written to concurrently. We use [double-checked locking](https://en.wikipedia.org/wiki/Double-checked_locking) where exclusive access to the datastructure for writing is required:
+
+```go
+func ensureValue(key string, f func() string) string {
+    // fast path
+    mutex.RLock()
+    value, ok := m[key]
+    mutex.RUnlock()
+    if !ok {
+        return value
+    }
+
+    // slow path
+    mutex.Lock()
+    defer mutex.Unlock()
+
+    if value, ok := m[key]; ok{
+        return value
+    }
+
+    // note: f() can be moved outside of critical
+    // section if it's expensive/side-effect free
+    value = f()
+    m[key] = value
+    return value
+}
+```
+
+This pattern first acquires a cheap read lock on `mutex` and checks the likely condition that the data has already been cached in the shared map `m`. If the value exists, it can be returned. Otherwise, a value needs to be generated and written to the shared map for later readers. In order to handle the race condition between releasing the read lock and acquiring the write lock, we check for the value in the map again after exclusive access has been acquired.
+
+Each shared datastructure has its own mutex so that contention from one resource does not bleed into a less-contended resource. However, it's easy to take this pattern too far. We have a shared two-layer hierarchical map whose access is protected by a [striped mutex](https://guava.dev/releases/19.0/api/docs/com/google/common/util/concurrent/Striped.html). A striped mutex is basically an array of shared mutexes, where a resource `r` is locked if the mutex at index `hash(r)` is locked (modulo the configured number of mutexes). This is a good middle-ground between having one mutex for a highly contended resource, and having `n` distinct mutexes for a map with `n` keys. The former will decrease the throughput of of the indexer as every routine is waiting on the same lock, and the latter will blow up memory usage, as `n` can be very large even for relatively small code bases.
+
+##### You have to be sure that there are no hidden data dependencies
```

Comment:
Is this is a problem introduced by parallelization, or a problem that is independent of parallelization? It isn't clear initially, but after reading through, it seems like it is related to parallelization, so mention that at the beginning as a segue. It may also make sense to group all the parallelization-related sections into a larger super-section titled "Parallelization" so this grouping is clear to the reader.